### PR TITLE
MQTT plain text demo

### PR DESCRIPTION
**Pull request by leegeth**
*Description of changes:*
MQTT plain text demo
Demo for showing the use of MQTT APIs to establish an MQTT session,
subscribe to a topic, publish to a topic, receive incoming publishes,
unsubscribe from a topic and disconnect the MQTT session.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: demos/mqtt/mqtt_demo_plaintext/mqtt_demo_plaintext.c [hunk 1]
```diff
@@ -148,135 +421,248 @@ static int connectToServer( const char * pServer, uint16_t port )
...
+    /* Reduce it from globalEntryTime so as to always return the
+     * elapsed time in the application. */
+    timeMs = ( uint32_t ) ( timeMs - globalEntryTimeMs );
+
+    return timeMs;
+}
+
+/*-----------------------------------------------------------*/
+
+static void handleIncomingPublish( MQTTPublishInfo_t * pPublishInfo,
```
[the remark this answers is not in the document]
Comment (cobusve) [reply demos/mqtt/mqtt_demo_plaintext/mqtt_demo_plaintext.c:486]: I think it is important not to filterin this simple demo, that should be a more advanced one. Topic filters are not part of the MQTT spec, but in practice if you have multiple applications the only way to route data is based on the topic or the content and the topic is definitely recommended.

**Comment by cobusve** [comment]
I really like the shape of the demo. One thing I think we can improve is being more explicit about the split between the different components. I think if I had to draw the boxes for the demo documentation I would draw a box for MQTT, one for the Transport interface, one for the Main application. I would suggest that splitting the demo into 3 or more files, each with a single responsibility will make this demo much easier to understand.

This will allow someone to start at the top of main and see a very simple and easy to understand implementation and then peel back the layers by looking into the other files.

The last time I did a demo like this my requirement was that main.c for the demo had to fit on the a typical PC screen all at the same time. This re-inforces the simplicity and ease of use. 

Even if this change is not made I can approve this, please let me know if you want to make the improvement or go ahead as, is.

**Comment by leegeth** [comment]
> to start at the top of main and see a very simple and easy to unde



> I really like the shape of the demo. One thing I think we can improve is being more explicit about the split between the different components. I think if I had to draw the boxes for the demo documentation I would draw a box for MQTT, one for the Transport interface, one for the Main application. I would suggest that splitting the demo into 3 or more files, each with a single responsibility will make this demo much easier to understand.
> 
> This will allow someone to start at the top of main and see a very simple and easy to understand implementation and then peel back the layers by looking into the other files.
> 
> The last time I did a demo like this my requirement was that main.c for the demo had to fit on the a typical PC screen all at the same time. This re-inforces the simplicity and ease of use.
> 
> Even if this change is not made I can approve this, please let me know if you want to make the improvement or go ahead as, is.

Thanks Cobus. I like that proposal and we were planning to do a similar restructuring of demos to separate out the platform dependent layers. I prefer the layers that you proposed to separate out even MQTT logic from main application. I think I will go ahead with this structure for now and come back to it to rearrange the demo in different layers.

File: demos/mqtt/mqtt_demo_plaintext/mqtt_demo_plaintext.c [hunk 2]
```diff
@@ -60,30 +84,237 @@
...
-    int status, tcpSocket = -1;
+    int status = EXIT_SUCCESS;
     struct addrinfo * pListHead = NULL, * pIndex;
     struct sockaddr * pServerInfo;
     uint16_t netPort = htons( port );
     socklen_t serverInfoLength;
+    struct timeval transportTimeout;
 
     /* Perform a DNS lookup on the given host name. */
     status = getaddrinfo( pServer, NULL, NULL, &pListHead );
```
[the remark this answers is not in the document]
Comment (leegeth) [reply demos/mqtt/mqtt_demo_plaintext/mqtt_demo_plaintext.c:320]: Yea. Thanks

File: demos/mqtt/mqtt_demo_plaintext/mqtt_demo_plaintext.c [hunk 3]
```diff
@@ -148,135 +421,248 @@ static int connectToServer( const char * pServer, uint16_t port )
...
- *
- * @return Number of bytes sent; negative value on error.
- */
-static int32_t transportSend( MQTTNetworkContext_t tcpSocket, const void * pMessage, size_t bytesToSend )
+static int32_t transportSend( MQTTNetworkContext_t tcpSocket,
+                              const void * pMessage,
+                              size_t bytesToSend )
 {
-    return ( int32_t ) send( tcpSocket, pMessage, bytesToSend, 0 );
+    return ( int32_t ) send( ( int ) tcpSocket, pMessage, bytesToSend, 0 );
```
[the remark this answers is not in the document]
Comment (leegeth) [reply demos/mqtt/mqtt_demo_plaintext/mqtt_demo_plaintext.c:428]: Sure. Thanks for pointing out.

File: demos/mqtt/mqtt_demo_plaintext/mqtt_demo_plaintext.c [hunk 4]
```diff
@@ -60,30 +84,237 @@
...
+
+/**
+ * @brief Delay between MQTT publishes in seconds.
+ */
+#define DELAY_BETWEEN_PUBLISHES_SECONDS     ( 1U )
+
+/**
+ * @brief Transport timeout in milliseconds for transport send and receive.
+ */
+#define TRANSPORT_SEND_RECV_TIMEOUT_MS      ( 20 )
```
[the remark this answers is not in the document]
Comment (leegeth) [reply demos/mqtt/mqtt_demo_plaintext/mqtt_demo_plaintext.c:138]: No. Timeval. tv_usec is signed. This macro is used for to assign value to Timeval. tv_usec

File: demos/mqtt/mqtt_demo_plaintext/mqtt_demo_plaintext.c [hunk 5]
```diff
@@ -148,135 +421,248 @@ static int connectToServer( const char * pServer, uint16_t port )
...
+
+        case MQTT_PACKET_TYPE_UNSUBACK:
+            LogInfoWithArgs( "Unsubscribed from the topic %.*s.",
+                             MQTT_EXAMPLE_TOPIC_LENGTH,
+                             MQTT_EXAMPLE_TOPIC );
+            /* Make sure ACK packet identifier matches with Request packet identifier. */
+            assert( globalUnsubscribePacketIdentifier == packetIdentifier );
+            break;
+
+        case MQTT_PACKET_TYPE_PINGRESP:
```
[the remark this answers is not in the document]
Comment (leegeth) [reply demos/mqtt/mqtt_demo_plaintext/mqtt_demo_plaintext.c:560]: MQTT_ProcessLoop sends the pingresp incoming packet to the application using the appcallback. https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/development/libraries/standard/mqtt/src/mqtt.c#L686
So thought of just logging the reception of PINGRESP here.

File: demos/mqtt/mqtt_demo_plaintext/mqtt_demo_plaintext.c [hunk 6]
```diff
@@ -148,135 +421,248 @@ static int connectToServer( const char * pServer, uint16_t port )
...
+    /* Get the MONOTONIC time. */
+    clock_gettime( CLOCK_MONOTONIC, &timeSpec );
+
+    /* Calculate the milliseconds from timespec. */
+    timeMs = ( uint32_t ) ( timeSpec.tv_sec * 1000 )
+             + ( uint32_t ) ( timeSpec.tv_nsec / ( 1000 * 1000 ) );
+
+    /* Reduce it from globalEntryTime so as to always return the
+     * elapsed time in the application. */
+    timeMs = ( uint32_t ) ( timeMs - globalEntryTimeMs );
```
[the remark this answers is not in the document]
Comment (leegeth) [reply demos/mqtt/mqtt_demo_plaintext/mqtt_demo_plaintext.c:479]: Yea. Will fix the comments. Thanks

File: demos/mqtt/mqtt_demo_plaintext/mqtt_demo_plaintext.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (abhidixi11) [reply demos/mqtt/mqtt_demo_plaintext/mqtt_demo_plaintext.c]: OK I missed that part, so yes, it probably needs to be there, but for process loop with keep alive it needs to be handled internally completely (send and receive ping). Looking at the code it feels like we have sort of hybrid Process loop. We can fix it in a separate PR.

**Pull request merged**